File: python/firstBadVersion.py
# https://leetcode.com/problems/first-bad-version/

# Example of a possible answer (>=1)
ANSWER = 4


def isBadVersion(version: int) -> bool:
    """Assume that this function is provided."""
    return version >= ANSWER


class Solution:
    def firstBadVersion(self, n: int) -> int:
        """
        Time: O(log(n))
        Space: O(1)
        """
        lo, hi = 1, n
        while lo < hi:
            mid = lo + (hi - lo) // 2

            # Case 1: The first bad version could be right here or
            # it could be earlier
            if isBadVersion(mid):
                hi = mid
            # Case 2: The first bad version must be to the right of
            # this particular version, so let's go up
            else:
                lo = mid + 1

        # By this time, we have the first bad version since we have
        # effectively converged on a single point
        return lo

File: java/FirstBadVersion.java
// https://leetcode.com/problems/first-bad-version/

import helper.VersionControl;

public class FirstBadVersion extends VersionControl {
    /**
     * Identify first bad version. Assume that all versions after a
     * bad version are also bad.
     *
     * <p>Complexity:
     *
     * <ul>
     *     <li>Time: O(log(n))</li>
     *     <li>Space: O(1)</li>
     * </ul>
     */
    public int firstBadVersion(int n) {
        int lo = 1, hi = n, first = n;

        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (isBadVersion(mid)) { // look down for earlier versions
                first = mid;
                hi = mid - 1;
            } else { // look up for newer versions
                lo = mid + 1;
            }
        }

        return first;
    }
}
